### ISSUE TITLE
Resource Pool Thread Killed

### ISSUE THREAD

**Issue by samdphillips** [Bogdanp/racket-http-easy#12]
I wrote a little proxy that reads a json off of a remote server and rewrites it into Prometheus exposition format.  If the remote server is unreachable and causing http-easy to raise an tcp-connect exception, the web server captured it and in turn killed the resource pool manager.  After that all subsequent requests fail because the resource-pool manager thread is dead.

```
resource-pool: leased idle resource #<http-conn>
http-easy: response: #"HTTP/1.1 200 OK"
resource-pool: adding #<http-conn> to idle set
resource-pool: leased idle resource #<http-conn>
http-easy: connection failed: tcp-connect: connection failed
  hostname: alexandria.local
  port number: 8080
  system error: Connection refused; errno=61
resource-pool: adding #<http-conn> to idle set
sensor: get-sensor-data: an error occurred
tcp-connect: connection failed
  hostname: alexandria.local
  port number: 8080
  system error: Connection refused; errno=61
  context...:
   /Applications/Racket v8.0/collects/net/http-client.rkt:67:0: http-conn-open!
   /Applications/Racket v8.0/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /Users/sphillips/Library/Racket/8.0/pkgs/http-easy/http-easy/private/session.rkt:300:0
   /Applications/Racket v8.0/collects/racket/contract/private/arrow-higher-order.rkt:379:33
   /Users/sphillips/Library/Racket/8.0/pkgs/http-easy/http-easy/private/pool.rkt:72:15

resource-pool: leased idle resource #<http-conn>
sensor: get-sensor-data: an error occurred
thread-send: target thread is not running
  context...:
   /Users/sphillips/Library/Racket/8.0/pkgs/resource-pool-lib/pool.rkt:173:0: pool-take!
   /Applications/Racket v8.0/collects/racket/contract/private/arrow-higher-order.rkt:375:33: ...row-higher-order.rkt:375:33
   /Users/sphillips/Library/Racket/8.0/pkgs/http-easy/http-easy/private/pool.rkt:60:0: pool-lease
   /Applications/Racket v8.0/collects/racket/contract/private/arrow-higher-order.rkt:379:33: ...row-higher-order.rkt:379:33
   /Users/sphillips/Library/Racket/8.0/pkgs/http-easy/http-easy/private/session.rkt:144:2: request
   /Users/sphillips/projects/racket-dev/rpi/sensor-scrape/sensor.rkt:20:2
   /Users/sphillips/projects/racket-dev/rpi/sensor-scrape/main.rkt:9:0: get-sensor-metrics
   /Users/sphillips/projects/racket-dev/rpi/sensor-scrape/endpoint.rkt:22:5: temp9
   /Applications/Racket v8.0/share/pkgs/web-server-lib/web-server/http/response.rkt:171:7

sensor: get-sensor-data: an error occurred
thread-send: target thread is not running
  context...:
   /Users/sphillips/Library/Racket/8.0/pkgs/resource-pool-lib/pool.rkt:173:0: pool-take!
   /Applications/Racket v8.0/collects/racket/contract/private/arrow-higher-order.rkt:375:33: ...row-higher-order.rkt:375:33
   /Users/sphillips/Library/Racket/8.0/pkgs/http-easy/http-easy/private/pool.rkt:60:0: pool-lease
   /Applications/Racket v8.0/collects/racket/contract/private/arrow-higher-order.rkt:379:33: ...row-higher-order.rkt:379:33
   /Users/sphillips/Library/Racket/8.0/pkgs/http-easy/http-easy/private/session.rkt:144:2: request
   /Users/sphillips/projects/racket-dev/rpi/sensor-scrape/sensor.rkt:20:2
   /Users/sphillips/projects/racket-dev/rpi/sensor-scrape/main.rkt:9:0: get-sensor-metrics
   /Users/sphillips/projects/racket-dev/rpi/sensor-scrape/endpoint.rkt:22:5: temp9
   /Applications/Racket v8.0/share/pkgs/web-server-lib/web-server/http/response.rkt:171:7
```

The original code I wrote that caused this had other problems so I ended up writing it in a way where this error no longer occurs.  [Here is the code base](https://github.com/samdphillips/sensor-scrape).  It's hopefully not too much to digest.  The commented out function at https://github.com/samdphillips/sensor-scrape/blob/main/endpoint.rkt#L26 is the version that would consistently fail.

**Comment by jackfirth** [comment]
Just shotgunning a guess from the peanut gallery, but maybe the resource-pool manager thread should be created with `thread/suspend-to-kill` instead of `thread` and code interacting with the manager thread should call `thread-resume` on the manager thread before trying to send or receive messages to it. I'm basing this off the [Kill-Safe Synchronization Abstractions](https://www.cs.utah.edu/plt/kill-safe/) paper.

**Comment by Bogdanp** [comment]
@jackfirth I think you're right. Pools are created lazily and the web server creates a new custodian for every request so the first custodian to make a request ends up being the owner of the pool's manager thread. That custodian then shuts down and takes the manager down with it so subsequent requests fail.  The following snippet reproduces the issue:

```racket
#lang racket/base

(require net/http-easy)

(define (test)
  (define cust (make-custodian))
  (parameterize ([current-custodian cust])
    (thread-wait
     (thread
      (lambda ()
        (displayln (response-body (get "http://example.com"))))))
    (custodian-shutdown-all cust)))

(test)
(test)
```

Making the thread suspendable and adding `thread-resume` in the right places is the right fix, but in doing so I ended up running into https://github.com/racket/racket/issues/3813 so I'll have to make more changes to work around that problem, assuming the fix will only go out in 8.2.

**Comment by samdphillips** [comment]
Thanks @Bogdanp !
